[#92502564] Add GCE configs

dcarley opened the pull request:
This is a mirror of AWS.

There are a few things outstanding, but this branch has already dragged on a long time, accumulated a lot of commits, and needs constant rebasing. So I'm going to raise this for review now and we can fix the other things within the same story.
- API isn't available over SSL because we only have access to L4 load balancing and not L7 to terminate SSL
- I haven't been able to verify whether the router is working after making the health check changes. Might require more tweaks to firewalls

I have however been able to use the outputted IPs to provision the cluster with Ansible, login, add a platform and deploy an app.

jimconner committed:
Starting point for GCE provisioning work

jimconner committed:
Subnet definitions for GCE

We have not included the public2 and private2 subnets
as they are AWS VPC related requirements.
We have not added routes because we do not have a NAT box.

jimconner committed:
Added postgres server role manifest.

jimconner committed:
Removed ephemeral IP from postgres instance

- it should only be accessible internally - not from the Internet

jimconner committed:
Jumpbox instance required for ssh access to non-public VMs

dcarley committed:
Remove public/private subnet split in GCE

Copying the same architecture from our AWS configs isn't possible because
while AWS magically routes between different subnets belonging to the same
VPC, GCE does not. As described in:

https://cloud.google.com/compute/docs/networking#networks_1

> Any communication between instances in different networks, even within the
> same project, must be through external IP addresses.

This would mean a complicated setup whereby each subnet has it's own NAT and
we'd be charged for traffic between the two subnets. So we're flattening
these to a single "private" subnet which will have some instances that are
exposed by external IPs or load balancers.

We'll figure out whether we can make AWS match at a later date.

dcarley committed:
Replace GCE jumpbox with AWS-like NAT

In order to match the AWS configuration. Also applying the provisioner which
configures this machine to do forwarding.

The `can_ip_forward` option on GCE is equivalent to `source_dest_check` on
AWS.

dcarley committed:
Adapt NAT provisioner for Ubuntu on GCE

These commands were previously for Amazon Linux which is a fork of CentOS.

dcarley committed:
Add route and tags for private instances

So that instances that don't have a public IP address can access the
internet through our NAT instance. These tags will represent the same as our
subnets in AWS.

dcarley committed:
Add docker-registry instance for GCE

Copying AWS. Not using a larger instance size because standard is close to
medium and should be sufficient.

dcarley committed:
Allow SSH to NAT instance in GCE.

dcarley committed:
Add GCE firewall rules for Gandalf and web

Matching AWS. The instances with these tags will be created in a follow
commit.

dcarley committed:
Copy app-servers for GCE

Contrary to the name this contains instances and load balancers for:

- API
- Router
- Gandalf

We should split these out for both at a later date.

Not using auto scaling groups because they aren't supported by Terraform for
GCE. We're currently unsure about whether we're going to continue using auto
scaling groups for AWS because we haven't integrated them with Ansible's
config management.

Any tags necessary for firewall rules will be fixed in a following commit.

dcarley committed:
Add firewall rule for GCE internal routing

This serves the same effect as the "default" rule in AWS which is associated
with all instances in the VPC.

dcarley committed:
Assign firewall rule to Gandalf instance in GCE

So that these can be accessed externally via the forwarding rule.

dcarley committed:
Assign firewall rule to Router+App in GCE

So that these can be accessed externally via the forwarding rule.

dcarley committed:
Add shell script to wrap GCE DNS

Terraform doesn't have a resource type for Google's Cloud DNS service. So
for the moment we need to call out to the `gcloud` command line utility.

However it doesn't itself provide a simple way to ensure that a record is
set to the value that we want. So we need to wrap it with a shell script
that deletes the existing record of the same name, regardless of it's
attributes, if it already exists.

dcarley committed:
Register DNS records in GCE

Using the wrapper script from the previous commit.

The `tsuru2` zone has been created by hand, like AWS, because it needs to be
delegated from the parent zone which is hosted by Dyn.

Handles the follow instances:

- API LB
- Router LB
- Docker registry internal IP
- Gandalf external IP
- NAT external IP

dcarley committed:
Balance GCE instances across zones

So that we are fault tolerant. This has more-or-less the same effect as
supplying a list of zones to the AWS provider and it using round-robin to
choose one.

The `split()` could be removed when Terraform has the ability to store
arrays in variables: hashicorp/terraform#57

dcarley committed:
Standardise on 2sp idents for all GCE configs

This matches *most* of the configs I've been writing (because it's how my
editor is configured) and matches the AWS configs.

This commit is best viewed with git's `-b` option or GitHub's `?w=1` option
- which should show no changes at all.

dcarley committed:
Remove `depends_on` for GCE network

This was carried over from our initial testing in the `gce-terraform` branch
where we were having problems with instances being created before the
networks that they depend on.

I've since better understood how Terraform does implicit relationships -
they are calculated based on variables that are constructed from other
resources. In our original test we were referring to a variable name with
`network = "${var.gce_net_name}"` whereas we are now referring to the actual
network resource, so Terraform should do the right thing.

I've confirmed this by using `terraform graph` before and after.

jimconner committed:
Added timstamp+uuid transaction-file name to prevent request clash

We noticed a problem when terraform was attempting to change two
records in parallel, whereby the gcloud command was trying to use
the transaction.yaml file for both operations simultaneously.
The gcloud command can accept a --transaction-file flag, so by
using unique transaction file names we prevent the clash.

jimconner committed:
Check error message response from gcloud command line

If there is already a transaction pending, then the current
one will fail... but it doesn't tell us this until we try
to execute the transaction.

We parse the error message back from the gcloud command to
look for code: 412 messages (Prerequisites not met) and
should we find that message, we have to abort the current
transaction and create a new one from the start.

jimconner committed:
First draft of outputs.tf

Using AWS version as a template
Updated field names appropriately for GCE provider

dcarley committed:
Indent loop in ensure_gce_dns script

To make it clear that all of this is inside a loop.

dcarley committed:
Copying SSL proxy from AWS to GCE

This mimics commit c58815d91beb10648c05268eeb440efdeef48a46 to GCE. The
major differences are, we don't add:

- add an internal LB for the router because GCE has no concept of internal
  LBs
- add new subnets because we're not using subnets on GCE as described in an
  earlier commit of this branch
- new security groups (firewall rules) because we already allow TCP/443 to
  instances that are tagged `web`

We are using a static IP address instead of ephemeral because we need to
forward both port 80 and 443, which we can't do with a forwarding rule
without specifying a range of ports that would be too open. We tried
specifying `ip_address` of one forwarding rule to another but it complained
about:

    Invalid value for field 'resource.natIP': 'natIP/104.155.25.127'. Resource was not found.

dcarley committed:
Add outputs for registry/postgres/sslproxy/router

So that we know the private IP addresses to add to Ansible's inventory file.
I've updated both GCE and AWS so that they have parity.

Except for the sslproxy instances on AWS, because Terraform doesn't know
about the instances which have been created by the auto-scaling group.

dcarley committed:
Group outputs for app servers on GCE

Using `*` and `join()` to list all addresses.

dcarley committed:
Only add single API instance to GCE LB

In order to match AWS. Only one of our "app" servers is running the API.
This will be addressed in story #91206264.

dcarley committed:
Use separate health check for API

This needs to be a different port and path to our "default" HTTP health
check. The path is important because tsuru's API responds to `/` with a 404
which causes the check to be unhealthy. The `/info` endpoint seems like a
reasonably lightweight thing to hit regularly and returns 200:

http://docs.tsuru.io/en/stable/reference/api.html#metadata

dcarley committed:
Split out Docker node for GCE

This copies the following commits from AWS:

- d4b9f2db91c8ee8353a6be2a74a763667af9aa74
- ce2c389e9cb6f9d1465e95c921dd6ca6074f377f

I've also removed the `web` tag from the Docker node because it could allow
them to receive public traffic from the Internet which isn't desirable
because they are not addressed directly.

dcarley committed:
Split app-servers.tf to separate files for GCE

This copies commit 4073ea4e634df3f28d19cb16edce0d8f14a1b15f from AWS. It's a
noop when running `terraform plan` against an existing install.

dcarley committed:
Split out separate API servers on GCE

This copies the following commits from AWS:

- 71d02ce3c87a9de61930068b51b444220e044a49
- 20c58fab777333e946e46e7a0a04472e7efb4f91

With the differences that:

- we're not using auto-scaling groups because they're not supported by
  Terraform on GCE
- I've removed the `web` tag from the old "app" server because it should no
  longer be receiving public traffic
- I've added a new output because we need the IPs of these new machines to
  put into Ansible's inventory (Terraform doesn't know the IPs of the ASG in
  AWS)

dcarley committed:
Remove health check from Router LB in GCE

The routers are difficult to health check because they don't respond to
requests themselves. Everything gets proxied to something else below. For a
brand new install it doesn't have any routes configured and for an existing
setup we don't have an app that we can rely on being there.

GCE only supports HTTP and HTTPS health checks on "forwarding rules". We
might have more options for "load balancers" but they aren't currently
supported by Terraform. So remove the health checks for now and always send
traffic to members.

dcarley committed:
Increase size of docker-registry disk on GCE

Copying commits from AWS:

- 8be049ddd54607d12ae42938136375b2b68bd068 (merge - see comment)
- 769e4d5ba6ffec27413ae0398737eebec6fe1349

mikkoc commented on aws/outputs.tf:
```diff
@@ -6,6 +6,10 @@ output "docker.private.ip" {
   value = "${aws_instance.tsuru-docker.private_ip}"
 }
 
+output "docker-registry.private_ip" {
```
Did this slip in or was it intended?

dcarley replied:
The commit f318c0f231086dc7e122ff504607c71dc9fca7b9 that it belongs to explains why.

mikkoc commented:
If it's too much time to change that, we can merge and tackle it with a separate PR! :+1:

dcarley committed:
Remove output for Router in AWS

These are in an auto-scaling group so Terraform doesn't know about the
instances.

annashipman commented:
I agree that the comment raised in a07cf55f should go on our tech debt list. As to how we'd fix it or whether we'd get around to it, we don't have to decide that now, and if we're lucky it will get fixed by someone else :smile:.

Is the issue raised in 8d33f46d worth adding to tech debt also?

Other than that, :+1: for getting rid of long-running branches and the other issues that you and @mikkoc raise can be addressed in subsequent PRs.

What I'd like to see in the future is some shared code and the differences in a different place, as already this has diverged from AWS in quite a few very small ways. However, that is a story for another day.

I'm going to merge this now but let's agree what to do about the tech debt stories.

mikkoc commented:
Well done! :+1:

dcarley commented:
Thanks! Still not quite there, but a lot closer.
